pymongo/pool.py

```diff
@@ -1257,8 +1257,17 @@ def _get_socket(self, all_credentials):
         if not self._socket_semaphore.acquire(
                 True, self.opts.wait_queue_timeout):
             self._raise_wait_queue_timeout()
-        with self.lock:
-            self.active_sockets += 1
+
+        try:
+            with self.lock:
```

Comment:
If gevent raises on the acquire here, self._socket_semaphore's lock will be permanently decremented.

Reply:
I think it would be simpler to move the `active_sockets` modification down below inside the existing try/except:
```python
# We've now acquired the semaphore and must release it on error.
sock_info = None
try:
    with self.lock:
        self.active_sockets += 1
    while sock_info is None:...
```

Reply:
I thought of that, however the second try/except block will decrement the active sockets if an exception is raised.

If you raise within the first with self.lock, you will not have first incremented active_sockets, so you will end up permanently decreasing active_sockets.

Reply:
Good point. What about adding a flag:
```python
# We've now acquired the semaphore and must release it on error.
sock_info = None
incremented = False
try:
    with self.lock:
        self.active_sockets += 1
        incremented = True
    while sock_info is None:...
except Exception:
    if sock_info:
        # We checked out a socket but authentication failed.	                
        sock_info.close_socket(ConnectionClosedReason.ERROR)
    self._socket_semaphore.release()
    if incremented:
        with self.lock:
            self.active_sockets -= 1
```

Reply:
That would work... Should I add that?

Reply:
Updated, seems to still work